src/main/java/reposense/report/RepoCloner.java

```diff
@@ -47,7 +63,41 @@
      */
     public void cloneBare(RepoConfiguration config) {
         configs[currentIndex] = config;
-        isCurrentRepoCloned = spawnCloneProcess(config);
+
+        if (!config.isShallowCloningPerformed()) {
+            isCurrentRepoCloned = spawnCloneProcess(config);
+        } else {
+            boolean didShallowPartialCloneSucceed = spawnShallowPartialCloneProcess(config);
+            String shallowPartialBareRoot = FileUtil.getShallowPartialBareRepoPath(config).toString();
+
+            if (!didShallowPartialCloneSucceed || GitRevList.checkIsEmptyRepo(shallowPartialBareRoot)) {
+                isCurrentRepoCloned = spawnCloneProcess(config);
+                return;
+            }
+
+            List<String> graftedCommits = GitRevList.getRootCommits(shallowPartialBareRoot);
+            List<String> graftedCommitParents = GitCatFile.getParentsOfCommits(shallowPartialBareRoot, graftedCommits);
+
+            boolean didPartialCloneSucceed = spawnPartialCloneProcess(config);
+            if (!didPartialCloneSucceed) {
+                isCurrentRepoCloned = spawnCloneProcess(config);
+                return;
+            }
+
+            String partialBareRoot = FileUtil.getPartialBareRepoPath(config).toString();
+            Date sinceDate;
+            try {
+                List<String> distinctParents = graftedCommitParents.stream().distinct().collect(Collectors.toList());
+                sinceDate = GitShow.getEarliestCommitDate(partialBareRoot, distinctParents);
```

Comment:
Just a question, will just check the date one by one be very time-consuming? Or the number of parents will be very limited given that the date range given is not so much?

Reply:
The number of parents is quite limited. For the cases I have manually checked, the number of grafted commits has been < 5. So even if all of them are merge commits with 2 parents, the number of parents to check is at most 10.